### ISSUE TITLE
Onbloc x Gno Core Workshop

### ISSUE THREAD

**Issue by dongwon8247** [gnolang/hackerspace#61]
The purpose of this workshop is to share the improvement areas of Gno that Onbloc finds necessary for upcoming testnet4 & mainnet while working on [GnoSwap Contracts](https://github.com/gnoswap-labs/gnoswap), [GnoSwap Interface](https://github.com/gnoswap-labs/gnoswap-interface), [Adena Wallet](https://github.com/onbloc/adena-wallet), and [GnoScan](https://github.com/onbloc/gnoscan). As one of the active core partners of Gno, Onbloc wants to contribute to defining important tasks for the upcoming testnet4 launch by sharing our inputs and getting assigned to some of the core tasks for the coordinated efforts.

## Onbloc’s focus right now:
- **GnoSwap Open Beta:** The official open beta of GnoSwap on testnet4 (and Portal Loop) where users can try it. This beta version will include enhanced Swap router APIs, an automated GRC20 registry, and better UX/UI for pool-related actions. Adding marketing efforts too to raise Gno’s awareness in the crypto community. Aiming to launch by the end of April 2024.
- **Adena:** Working on v1.10.0, which includes an in-app faucet, pop-up window, injection improvement, and other minor bugs.
- **Onbloc APIs Improvements**: Replace our current indexer with the Gno tx-indexer to support custom RPC networks in Adena and GnoScan.

### 1. JSON
- Outline: JSON Encoder/Decoder with state machine.
- Current Progress: 
   - Finish for implementation ([#1415](https://github.com/gnolang/gno/pull/1415)), but it contains some inefficient workaround now that can be improved on a high level.
   - Guilhem proposed another PR ([#1776](https://github.com/gnolang/gno/pull/1776)), and Byeongjun thinks this would be a more efficient and better solution, but based on his work/research on JSON in #1415, he wants to modify or add features in #1776.
- Conclusion
   - Both #1415, #1776 to be reviewed by Milos and Petar.

### 2. bigint (u256)
- Outline: Packages for arbitrary precision for numbers outside of uint64 coverage in Go (Gno).
- Background
   - GnoSwap (or any other dApps that require complex calculations) requires support for arbitrary precision.
   - Attempt #1 bigint https://github.com/gnolang/gno/pull/764 → [Possible memory issues raised by Morgan](https://github.com/gnolang/gno/pull/764#issuecomment-1545363203)
   - Attempt #2 uint256/int256 ([#1778](https://github.com/gnolang/gno/pull/1778/files)) https://github.com/gnoswap-labs/gnoswap/tree/main/_lib_package/big 
   - Porting References: [uint256](https://pkg.go.dev/github.com/holiman/uint256), [int256](https://pkg.go.dev/github.com/mempooler/int256).
- Current Progress: 
   - Implemented and applied in GnoSwap Contracts
https://github.com/gnoswap-labs/gnoswap/tree/main/_lib_package/big 
   - All test cases passing but,
   - Failing txtar tests  Presumably a GnoVM Bug; unlikely a bug with u256/i256), and debugging to find problems.
- Discussions
   - WIP https://github.com/gnolang/gno/pull/1778 → Need a review by the Core team once done.

### 3. Emit/event
- Outline: modified previous PR to fit into current gno ([#1653](https://github.com/gnolang/gno/pull/1653))
- Background
   - Now relying on our own Sync Program & DB (Onbloc APIs) → Resource intensive & lack of real-time data support
   - Many previous attempts
       - https://github.com/gnolang/gno/issues/575
      - https://github.com/gnolang/gno/pull/853 
      - https://github.com/gnolang/gno/issues/975 
- Current Progress: Now blocked on code gen
- Discussions
   - Critical for testnet4.
   - Minor modifications/fixes to be made by Byeongjun. Then move on with reviews by Milos.

### 4. Tx-indexer
- Outline: 
   - Transaction Raw Content data parsing and schema-defining
   - Transaction filtering with GraphQL
   - Transaction list, Realm Package list (The GRC20 token list requires extra work from the implementer’s side)
- Current Progress: Being mainly developed by Milos and Antonio, and Byeungjun and Jinwoo are looking into contributing it.
- Discussions
   - https://github.com/gnolang/tx-indexer/pull/5 → To be reviewed by Milos.
   - Open an issue with features needed for Adena and Gnoscan by Jinwoo.

### 5. Transaction Chaining
- Outline: Running chained transactions with a single request. (such as the One-click staking feature of GnoSwap)
- Current Progress: Issue: https://github.com/gnolang/gno/issues/1508 raised but doing it with a workaround.
- Conclusion: Manfred recommends handling this by creating a function using a pattern for chaining conditional transactions rather than supporting this at the language level.
- Discussions
   - GnoSwap has implemented separate functions to handle these for now.
   - A nice-to-have feature for now.
   - TODO: take a look at how other blockchains solve these.

### 6. Contract Versioning
- Outline: Implementing versioning for realms to support updates or changes, while addressing challenges such as data loss, immutability, and data migration.
- Current Progress: Mainly by Hariom https://github.com/gnolang/gno/issues/694, https://github.com/gnolang/gno/pull/1631 
- Discussions
   - Required for Testnet4.
   - Awaiting review by Milos.

### 7. Multi-node Initiative
- Outline: Multinode testnet initiative to test and define node specs and make an official guide with the Gno team.
- Current Progress:
   - https://github.com/gnolang/hackerspace/issues/9
   - https://github.com/gnolang/gno/pull/871 
   - https://github.com/gnolang/gno/pull/889 
   - https://github.com/gnolang/hackerspace/tree/main/multinode-testnet 
- Discussions
   - Blocker: gno secrets PR https://github.com/gnolang/gno/pull/1593 
   - The PR to be worked on this week by Milos

### 8. Onchain Registry
- Outline: A callback pattern for using GRC20 tokens in realms without the import syntax.
- Current Progress:
   - Implemented to GnoSwap Contracts (Pool, Router, Staker) https://github.com/gnoswap-labs/gnoswap/blob/main/pool/token_register.gno
   - Example PR https://github.com/gnolang/gno/pull/1275
   - Manfred recommends using GRC20 objects. To do so, we need #1257 to be complete. https://github.com/gnolang/gno/pull/1257 
   - When a GRC20 token is deployed on the blockchain, it must be automatically registered. We are using the code in: https://github.com/gnoswap-labs/grc20-register 
- Discussion:
   - A nice-to-have feature for now.

### 9. DevX Improvements (setup, logs, debugging tool, tests etc)
- Outline: Some work being done by Byeungjun to improve the DevX.
   - Adding some functions for p/ufmt
   - Gno does not support os.Stdout, so I utilize println built-in function to access the stream buffer.
   - Add ParseFloat and add %f option for Sprintf (strconv?)
   - Fuzzing `testing.F`
   - Human generated test code has limitations.
   - Using genetic algorithm to manipulating the input data (mutation-based lexical fuzzing)
https://github.com/gnolang/gno/pull/1809
- Discussions
   - To be worked on by Onbloc

### 10. Unknown bugs
- unexpected object with id https://github.com/gnolang/gnochess/issues/97
   - Temporary patch: https://github.com/gnolang/gnochess/pull/129/files
   - Dragos patch: https://github.com/gnolang/gno/issues/1543 
   - struct with pointer field
   - save & restore state from map or avl => panic
- Discussions
   - To be reviewed by Petar.

### 11. Minor Pending PRs
- A PR to ensure that wugnot follows the specs of p/users and r/users after their recent changes. https://github.com/gnolang/gno/pull/1782
- A PR to add a test account in genesis for Adena’s faucet. https://github.com/gnolang/gno/pull/1808
